### ISSUE TITLE
Question: how to track path length during search?

### ISSUE THREAD

**Issue by ulysses4ever** [devonhollowood/search-algorithms#17]
This is a user question, and please let me know if there's a better place to ask this.

I'm using the `bfs` function, and I'd like to prune some of the `next` states based on how many steps we've taken. Is that possible? I tried `bfsM` with a simple state monad holding an integer, which I bumped every `next` call, but this gives me (I think) the number of states explored, not the number of states in the current path.

If this is not currently possible, would that be a useful addition to the package? Since the algorithm keeps track of the path internally, it shouldn't be hard to expose it (and, by extension, any feature of it, like the length) to the callbacks (probably, makes sense for `next` but not for `final` though).

The motivation is the following: I run `bfs` from several initial points in the space and want to find out which one of the initials gives me the shortest path. If there's a lot of initials (relative to the size of space) and/or they are close to each other, I'll keep observing the same states during various `bfs`'s. The idea is that I memoize the shortest paths for previously processed initials, and during next `bfs`s I can prune some states if the steps I took already plus the shortest path from that state is worse than the current record.

**Comment by ulysses4ever** [comment]
All right, I found a way to track depth: instead of an integer, I use  a map where I'm adding newly discovered nodes with depth of the current incremented by one.

And maybe extending the API as I suggest is tricky because it would expose the internals too much.

**Comment by devonhollowood** [comment]
Perfect, that sounds like exactly what I was going to recommend.

I wonder if it is possible to do this all in a single BFS call (with no memoization), by making your state somehow represent the BFS progress from all of the initial points. I.e., the first "round" of BFS would represent the initial points, the second "round" would represent all points within one step of the initial points, etc. The bookkeeping might be pretty messy, so your way is probably cleaner, but it's interesting to think about.

**Comment by ulysses4ever** [comment]
Yes, interesting question. On the other hand, when I solved the task I was solving (it's [AOC 2022: Day 12][aoc], btw), I learned that there was a better way: instead of running several bfs' to one final state, it's easier to run one bfs from that final state and see which of the initials is reached first. Of course, this hack only works because there's only one final state. In general, the question of how to run from several initials is a valid one.

[aoc]: https://adventofcode.com/2022/day/12
